### PR TITLE
[Part]Refactor Spiral and helix primitives

### PR DESCRIPTION
**Pull request by tomate44**
Use the same function to build spirals and helices.
Option to split these curves into any number of segments (no more restricted to 1-turn edges).

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, we ask you to conform to the following items. Pull requests which don't satisfy all the items below might be rejected. If you are in doubt with any of the items below, don't hesitate to ask for help in the [FreeCAD forum](https://forum.freecadweb.org/viewforum.php?f=10)!

- [x]  Your pull request is confined strictly to a single module. That is, all the files changed by your pull request are either in `App`, `Base`, `Gui` or one of the `Mod` subfolders. If you need to make changes in several locations, make several pull requests and wait for the first one to be merged before submitting the next ones
- [ ]  In case your pull request does more than just fixing small bugs, make sure you discussed your ideas with other developers on the FreeCAD forum
- [x]  Your branch is [rebased](https://git-scm.com/docs/git-rebase) on latest master `git pull --rebase upstream master`
- [ ]  All FreeCAD unit tests are confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ]  All commit messages are [well-written](https://chris.beams.io/posts/git-commit/) ex: `Fixes typo in Draft Move command text`
- [ ]  Your pull request is well written and has a good description, and its title starts with the module name, ex: `Draft: Fixed typos`
- [ ]  Commit messages include `issue #<id>` or `fixes #<id>` where `<id>` is the [FreeCAD bug tracker issue number](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) in case a particular commit solves or is related to an existing issue on the tracker. Ex: `Draft: fix typos - fixes #0004805`

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

### CODE REVIEW

File: src/Mod/Part/App/PrimitiveFeature.cpp [hunk 1]
```diff
@@ -850,16 +853,14 @@ App::DocumentObjectExecReturn *Helix::execute(void)
...
-//            this->Shape.setValue(helix.makeLongHelix(myPitch, myHeight, myRadius, myAngle, myLocalCS));
-//        else
-//            this->Shape.setValue(helix.makeHelix(myPitch, myHeight, myRadius, myAngle, myLocalCS, myStyle));
+        Standard_Real nbTurns = myHeight / myPitch;
+        Standard_Real mySegLen = SegmentLength.getValue();
+        
+        Standard_Real myRadiusTop = myRadius + myHeight * tan(Base::toRadians(myAngle));
+        TopoShape spirhelix;
+
+        TopoDS_Shape myHelix = spirhelix.makeSpiralHelix(myRadius, myRadiusTop, myHeight, nbTurns, mySegLen, myLocalCS);
```
Comment (davidosterberg) [inline src/Mod/Part/App/PrimitiveFeature.cpp:862]: Better write
`TopoDS_Shape myHelix = TopoShape().makeSpiralHelix(myRadius, ...`

File: src/Mod/Part/App/PrimitiveFeature.cpp [hunk 2]
```diff
@@ -913,30 +917,17 @@ App::DocumentObjectExecReturn *Spiral::execute(void)
...
-        BRepBuilderAPI_MakeFace mkFace(aPlane, -range, range, -range, range
-#if OCC_VERSION_HEX >= 0x060502
-        , Precision::Confusion()
-#endif
-        );
-        BRepProj_Projection proj(myHelix, mkFace.Face(), gp::DZ());
-        this->Shape.setValue(proj.Shape());
-
+        
+        TopoDS_Shape mySpiral = spirhelix.makeSpiralHelix(myRadius, myRadiusTop, 0, myNumRot, mySegLen, Standard_False);
```
Comment (davidosterberg) [inline src/Mod/Part/App/PrimitiveFeature.cpp:929]: Agan, better write
`TopoDS_Shape myHelix = TopoShape().makeSpiralHelix(myRadius, ...`

File: src/Mod/Part/App/TopoShape.cpp [hunk 3]
```diff
@@ -2503,6 +2503,62 @@ TopoDS_Shape TopoShape::makeLongHelix(Standard_Real pitch, Standard_Real height,
...
+    // A Bezier curve is used below, to get a periodic surface also for spirals.
+    TColgp_Array1OfPnt poles(1,2);
+    poles(1) = gp_Pnt(fabs(radiusbottom), 0, 0);
+    poles(2) = gp_Pnt(fabs(radiustop), 0, height);
+    Handle(Geom_BezierCurve) meridian;
+    meridian = new Geom_BezierCurve(poles);
+
+    gp_Ax1 axis(gp_Pnt(0.0,0.0,0.0) , gp::DZ());
+    Handle(Geom_Surface) surf;
+    surf = new Geom_SurfaceOfRevolution(meridian, axis);
```
Comment (davidosterberg) [inline src/Mod/Part/App/TopoShape.cpp:2532]: Where is this memory freed?

**Review by davidosterberg — CHANGES_REQUESTED** [review]
Code looks ok. Only a few comments.

**Comment by davidosterberg** [comment]
I think you should make this function available to the python interpreter as well. By adding it to `src/Mod/Part/App/AppPartPy.cpp`

File: src/Mod/Part/App/TopoShape.cpp [hunk 4]
```diff
@@ -2503,6 +2503,62 @@ TopoDS_Shape TopoShape::makeLongHelix(Standard_Real pitch, Standard_Real height,
     return TopoDS_Shape(std::move(wire));
 }
 
+TopoDS_Shape TopoShape::makeSpiralHelix(Standard_Real radiusbottom, Standard_Real radiustop,
+                                  Standard_Real height, Standard_Real nbturns,
+                                  Standard_Real breakperiod, Standard_Boolean leftHanded) const
+{
+    // 1000 periods is an OCCT limit. The 3D curve gets truncated
+    // if he 2D curve spans beyond this limit.
```
Comment (luzpaz) [inline src/Mod/Part/App/TopoShape.cpp:2511]: @tomate44 do you mind `s/he/the/` on this line?

File: src/Mod/Part/App/PrimitiveFeature.cpp [hunk 5]
```diff
@@ -879,12 +880,15 @@ Spiral::Spiral(void)
     Radius.setConstraints(&quantityRange);
     ADD_PROPERTY_TYPE(Rotations,(2.0),"Spiral",App::Prop_None,"The number of rotations");
     Rotations.setConstraints(&quantityRange);
+    ADD_PROPERTY_TYPE(SegmentLength,(1.0),"Spiral",App::Prop_None,"The number of turns of an spiral segment");
```
Comment (luzpaz) [inline src/Mod/Part/App/PrimitiveFeature.cpp:883]: Unrelated: do these strings that look like 'tooltips' get translated somewhere?

File: src/Mod/Part/App/TopoShape.cpp [hunk 6]
```diff
@@ -2503,6 +2503,62 @@ TopoDS_Shape TopoShape::makeLongHelix(Standard_Real pitch, Standard_Real height,
     return TopoDS_Shape(std::move(wire));
 }
 
+TopoDS_Shape TopoShape::makeSpiralHelix(Standard_Real radiusbottom, Standard_Real radiustop,
```
Comment (davidosterberg) [inline src/Mod/Part/App/TopoShape.cpp:2506]: makeLongHelix, returns a TopoShape. Which could be expected from a method in the TopoShape class. Personally I almost always want the TopoDS_Shape, so I don't really mind.

File: src/Mod/Part/App/PrimitiveFeature.cpp (hunk 5 above)
Comment (chennes) [reply src/Mod/Part/App/PrimitiveFeature.cpp:883]: Yep: they appear to be translated at point of use, e.g. https://github.com/FreeCAD/FreeCAD/blob/20c0f14efc5ebcd0722a9605140bb0686b3c2eaa/src/Gui/propertyeditor/PropertyItem.cpp#L335

File: src/Mod/Part/App/PrimitiveFeature.cpp (hunk 1 above)
Comment (tomate44) [reply src/Mod/Part/App/PrimitiveFeature.cpp:862]: OK.Thanks

File: src/Mod/Part/App/PrimitiveFeature.cpp (hunk 2 above)
Comment (tomate44) [reply src/Mod/Part/App/PrimitiveFeature.cpp:929]: OK

File: src/Mod/Part/App/TopoShape.cpp (hunk 3 above)
Comment (tomate44) [reply src/Mod/Part/App/TopoShape.cpp:2532]: ??? What are talking about ? You're talking to a python dev ! ;D
... a suggestion please ?

File: src/Mod/Part/App/TopoShape.cpp (hunk 4 above)
Comment (tomate44) [reply src/Mod/Part/App/TopoShape.cpp:2511]: Yes, "the", thanks

File: src/Mod/Part/App/PrimitiveFeature.cpp (hunk 5 above)
Comment (tomate44) [reply src/Mod/Part/App/PrimitiveFeature.cpp:883]: I don't know anything about translation mechanism ...

File: src/Mod/Part/App/TopoShape.cpp (hunk 6 above)
Comment (tomate44) [reply src/Mod/Part/App/TopoShape.cpp:2506]: No, makeHelix() and makeLongHelix() return TopoDS_Shape already.
I copy/pasted one of these two functions as a skeleton to make the new one.

File: src/Mod/Part/App/TopoShape.cpp (hunk 3 above)
Comment (tomate44) [reply src/Mod/Part/App/TopoShape.cpp:2532]: From what I see everywhere else in the file, there is no need to free this.
I think this is done by OCCT Handle

**Comment by davidosterberg** [comment]
Ok, I have compiled and tested the functionality by extending PartDesign Helix with this feature. It works brilliantly. With this, PD helix will be a complete generalization of the revolution tool: 

helix_spiral

File: src/Mod/Part/App/TopoShape.cpp (hunk 6 above)
Comment (davidosterberg) [reply src/Mod/Part/App/TopoShape.cpp:2506]: Ok, I checked again and you are right. Don't know what I was talking about.

File: src/Mod/Part/App/TopoShape.cpp (hunk 3 above)
Comment (davidosterberg) [reply src/Mod/Part/App/TopoShape.cpp:2532]: I think you are right. The handle seems to be a kind of smart pointer. It was simply a doubt I had, and this answers my question thanks.

**Comment by tomate44** [comment]
Great !

**Review by davidosterberg — APPROVED** [review]

**Comment by wwmayer** [comment]
* There is a typo in the tool tip text of the Part helix: "The number of turns of **an** helix segment" --> ... a helix ...
* There is a typo in the tool tip text of the Part spiral: "The number of turns of **an** spiral segment" --> .. a spiral ...

**Comment by tomate44** [comment]
Thanks. I am not quite sure with the wording of the tooltip, anyway.
I anyone has a better proposal ...

**Comment by davidosterberg** [comment]
Maybe "Number of turns per helix subdivision"

**Review by davidosterberg — APPROVED** [review]

**Comment by tomate44** [comment]
Thanks David and Werner for the review.
I have removed the [Do not merge] tag, as I think it is ready now.
Now I also need to work on makeHelix() and makeLonHelix() (both C++ and Python).
But I'd prefer to do another PR for that.
I'll ask for advice on the forum : https://forum.freecadweb.org/viewtopic.php?f=10&t=56343

**Comment by wwmayer** [comment]
Merged as 18f36c10, 6f9856b, 97543396c,  839b9638

I have squashed two commits.

**Pull request closed without merging**